sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextSourceTest.java

```diff
@@ -0,0 +1,143 @@
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class TextSourceTest {
+
+  @Test
+  public void testSubstringByteArrayOutputStreamSuccessful() throws IOException {
+    TextSource.SubstringByteArrayOutputStream output =
+        new TextSource.SubstringByteArrayOutputStream();
+    assertEquals("", output.toString(0, 0, StandardCharsets.UTF_8));
+
+    output.write("ABC".getBytes(StandardCharsets.UTF_8));
+    assertEquals("ABC", output.toString(0, 3, StandardCharsets.UTF_8));
+    assertEquals("AB", output.toString(0, 2, StandardCharsets.UTF_8));
+    assertEquals("BC", output.toString(1, 2, StandardCharsets.UTF_8));
+    assertEquals("", output.toString(3, 0, StandardCharsets.UTF_8));
+
+    output.write("DE".getBytes(StandardCharsets.UTF_8));
+    assertEquals("ABCDE", output.toString(0, 5, StandardCharsets.UTF_8));
+  }
+
+  @Test
+  public void testSubstringByteArrayIllegalArgumentException() throws IOException {
+    TextSource.SubstringByteArrayOutputStream output =
+        new TextSource.SubstringByteArrayOutputStream();
+    output.write("ABC".getBytes(StandardCharsets.UTF_8));
+
+    IllegalArgumentException exception;
+    exception =
+        assertThrows(
+            IllegalArgumentException.class, () -> output.toString(-1, 2, StandardCharsets.UTF_8));
+    assertThat(exception.getMessage(), containsString("offset is negative"));
+
+    exception =
+        assertThrows(
+            IllegalArgumentException.class, () -> output.toString(4, 0, StandardCharsets.UTF_8));
+    assertThat(exception.getMessage(), containsString("offset exceeds the buffer limit"));
+
+    exception =
+        assertThrows(
+            IllegalArgumentException.class, () -> output.toString(0, -1, StandardCharsets.UTF_8));
+    assertThat(exception.getMessage(), containsString("length is negative"));
+
+    exception =
+        assertThrows(
+            IllegalArgumentException.class, () -> output.toString(2, 2, StandardCharsets.UTF_8));
+    assertThat(exception.getMessage(), containsString("offset + length exceeds the buffer limit"));
+  }
+
+  @Test
+  public void testDelimiterFinder() {
+    assertEquals(Arrays.asList("A", "C"), split("AB", "AABC"));
+    assertEquals(Arrays.asList("A", "B", "C"), split("AB", "AABBABC"));
+
+    assertEquals(Arrays.asList("A", "C"), split("AAB", "AAABC"));
+    assertEquals(Arrays.asList("ABAB"), split("AAB", "ABAB"));
+    assertEquals(Arrays.asList("A", "A", "B"), split("AAB", "AAABAAABB"));
+
+    assertEquals(Arrays.asList("A", "B"), split("AABA", "AAABAB"));
+    assertEquals(Arrays.asList("AABBA"), split("AABA", "AABBA"));
+
+    assertEquals(Arrays.asList("", "D"), split("ABABC", "ABABCD"));
+    assertEquals(Arrays.asList("ABABAD"), split("ABABC", "ABABAD"));
+    assertEquals(Arrays.asList("ABABBD"), split("ABABC", "ABABBD"));
+    assertEquals(Arrays.asList("AB"), split("ABABC", "ABABABC"));
+
+    assertEquals(Arrays.asList(""), split("ABCABD", "ABCABD"));
```

Comment:
I don't see a test where there are several delimiters in a row, do we want multiple empty strings as result in that case?
ie
split("AA", "AAAAAA") ->  ["", "", ""] ?

Reply:
Done.